### ISSUE TITLE
Feature Request: Update key values & sync relations

### ISSUE THREAD

**Issue by razvanioan** [vuex-orm/vuex-orm#90]
First of all, really nice job here ;)

Being very new to Vue / Vuex and reading past few days the guides, examples and stuff (so my head is about to blow), documenting for a project that I need to do, I came across with the question how to handle DB structured data with the ecosystem, and then I've found your ORM plugin.

One thing I don't understand just yet, trying to combine this with a REST API.

When users will dynamically create new records, I will have to assign some local generated IDs and use those for the relationships (I am thinking to use negative integers, just to know that those are not real in the DB just yet) 

And whenever I sync data with the server I would like to update those "fake" IDs with real ones generated by DB server.

I have just tried to update the ID (key) field of some record like this:

```javascript
app.$store.dispatch('entities/users/update', {
  where: -2,
  data: { id: 9 }
})
```

Which updates the `id` value from the store object, but it doesn't change the store key as well, leaving it like this:

```javascript
{
  entities: {
    users: {
      '-2': { id: 9, name: 'Jane Doe' }
   }
}
```

Maybe there's something I miss in the documentation for this ...

And of course, a nice feature of some kind of `updateKey` method would be to also update all the related model records as well.

One way would be to use an extra `field` which corresponds with the real DB ID, but this would overcomplicate things to keep track of relationships on the client / server, continuously having to remap them.

**Comment by tjsteinhaus** [comment]
I'm actually working through this currently with a project. I have everything working for pulling in data from the REST API and storing it, along with making sure everything works for offline as well. I just started working on the piece to sync data back up to the server after being offline and I'll try and report back once I have more of a clear answer for you, but for now I'll try and explain how it currently works and what I plan on doing.

Whenever I am making updates while the user is online, I am always hitting the REST API with the new data before I add it to the Vuex ORM so I know what the ID's are without having to worry about making stuff up. However for offline mode, ID's are just generated when a new one gets added, but the way I believe I'm having it work is just to ignore the ID's when I push it back the server as just send a JSON object back with all the data and just parse through it. Basically the JSON object is just every piece of information that's setup with relationships.

**Comment by kiaking** [comment]
This is a kinda complicated issue that we should seriously discuss. We had few discussion on #78, but it's still not concluded.

> Which updates the id value from the store object, but it doesn't change the store key as well, leaving it like this:

This issue should be fixed. Honestly, I wasn't thinking about when people update its primary key, but when they do, it should update its object key as you would expect.

I've talked a bit about `replace` method to accomplish this on #78.

```js
// When having this kind of state.
{
  users: {
    data: {
      '1': { id: 1, name: 'John Doe' },
      '2': { id: 2, name: 'Jane Doe' }
    }
  }
}

// And replace some data.
store.dispatch('entities/users/replace', {
  where: 1,
  data: { id: 3, name: 'John Doe' }
})

// Then state becomes like this.
{
  users: {
    data: {
      '3': { id: 3, name: 'John Doe' }, // <- Replaced!
      '2': { id: 2, name: 'Jane Doe' }
    }
  }
}
```

However, I think we could just do this by update method so why would I want to introduce new method... I wasn't thinking enough I guess.

For the workaround, until we fix this issue, you could delete a record first and insert it as a new record : )

> And of course, a nice feature of some kind of updateKey method would be to also update all the related model records as well.

Now, this is some serious thinking required discussion. I need more thinking but updating related record is quite tough. Because if we have some related data that is belonging to the updated record and if we want it to be updated, then we have to go through all of the records and check if those record as belonging to the updated record.

This requires not only checking full records, but we have to also examine each entity if they have relations or not too. And those relations are bidirectional, so it is even more complicated.

But again, this sounds like really cool if we can accomplish it. I will think about this more but it would be very nice if I can get some ideas on this thread on actually how to implement such feature in a smartly way.

**Comment by razvanioan** [comment]
Hm, maybe I've missed the `replace` method in the guide, I'll give it a try !

It sounds better than deleting and reinserting the record, this might get a strange fuzzy reactive update in the views ...

As for updating related records, why to parse all the store data ?

For the moment, knowing the relationships myself, I will try to manually search the related records using `query` and use `update` only on them, as we don't have to update the other models keys, but only a parameter value !

This of course it would be nice to have automatically parsing the relationships definitions of the models, regarding the record we're updating the key !

**Comment by kiaking** [comment]
> Hm, maybe I've missed the replace method in the guide, I'll give it a try !

Oh no, sorry that `replace` method is not implemented yet! It was just a proposal on that thread. Sorry for the confusion. : (

> As for updating related records, why to parse all the store data?

Well because it is not always obvious that some model relates to the other.

For example, let's say a user defined following models.

```js
class User extends Model {
  static fields () {
    id: this.attr(null),
    name: this.attr('')
  }
}

class Post extends Model {
  static fields () {
    id: this.attr(null),
    user_id: this.attr(null),
    title: this.attr(''),
    user: this.belongsTo(User, 'user_id')
  }
}
```

Now, if we look at Post model, we can easily see that it belongs to a User, but User class has no idea that Post belongs to the User because User model doesn't have `hasMany` or `hasOne` relation to the Post.

So, if we have this kind of state.

```js
{
  users: {
    data: {
      '1': { id: 1, name: 'Jhon' }
    }
  },
  posts: {
    data: {
      '1': { id: 1, user_id: 1, title: '...' },
      '2': { id: 2, user_id: 1, title: '...' }
    }
  }
}
```

And if we were to update user id 1 to be the id of 2, then we have to update `user_id` of 2 posts record. But User model doesn't know that Post belongs to User, so we have to check Post model and see it has relation to the User or not. And then update all records that contain `user_id` of 2.

Maybe at first, we could add the limitation that if you want `update` method to update all related record, then you must define the relationship to the model that is being updated – in the example above, a user should define `hasMany` relation to the User model.

**Comment by razvanioan** [comment]
**Idea: Could an event bus be a solution here ?**

 * models having `id / keys` chnaged / updated to $emit such an event
 * related models to automatically listen for such events, and in case it applies to them, query it's data for the old keys and updated accordingly

**Comment by kiaking** [comment]
Hah, yeah that is a good idea! I'll give it a try. I think that works 🤝

**Comment by razvanioan** [comment]
Hi there @kiaking,

I've been _quiet_ for some time, as being busy with lots of projects & tasks, like we devs always are ;)
I've seen you've managed to complete some of the tasks on the **v1** roadmap, which is nice.

Is there any progress / news regarding this one in particular one ?
I'm close to the point I have to deal with this into my project, and have to find a working solution ;)

Maybe I'll try to implement myself one solution to this, event based or some other idea ...

Cheers !

**Comment by kiaking** [comment]
Hi! Sorry for making you wait! I really want to make this happen but I was having been very busy these days... I hope I can find time to tackle this one soon!

**Comment by razvanioan** [comment]
No man, don't get me wrong.
I was just curious if you get a chance to work at this point, you don't have to apologise

As promised, and being close to need this, I will let you know if I can tweak something out of it ;)

Thx,

**Comment by kiaking** [comment]
Thank you so much! It would be very helpful if you can find some solution to this...!

**Comment by razvanioan** [comment]
Looking a little bit into this, just by simple / manual updates into `model lifecycle hooks` I find it quite hard, as even if I manually update all `id` & `$id` fields using the framework itself, the `ids` of `entities` remain intact ...

Couldn't find the time (yet) for looking into some `smart` solution into the source code for this, and deleting & recreating new record with the new ID didn't seem natural (although a good workaround)

Because I wanted a really fast UX (and no dependency of server connectivity, at least until some point), I will also avoid AJAX inside hooks.

Thus, I will try a different approach:

- update all my model MySQL tables, and switch ID's from auto_increment INT to a UUID (stored as binary)
- update all the models to use ID's as strings
- I will generate UUIDs on client-side and just convert to binary on DB storage

At least until maybe there will be a nice solution for updating ID's on ORM

**Comment by kiaking** [comment]
@razvanioan Thank you so much for looking into this! I'll check this one as well. However, due to how Veux and Cue work, if we need to manipulate the key of the Object, we must recreate them. So I guess there's no other way than recreating at least the certain Object :/

**Comment by kiaking** [comment]
This is now implemented in Vuex ORM 0.31.3 🎉  However, it will not update related records, only the primary one.

The reason is, because usually when you update database field, it will not automatically update relationship ids by default. I'm thinking we should have another API, or some option to do that. We'll follow up that feature at different issue where we're discussing about adding relationship management #275.

https://github.com/vuex-orm/vuex-orm/releases/tag/v0.31.3

**Comment by d9k** [comment]
@kiaking, please implement cascade update for the foreign keys! :pray:
